Added function "removeGarbageParametersFromURL()"

aterehov opened the pull request:
Fixes: #6498

Added `removeGarbageParameterFromURL()` function that removes parameters of some minimal length that look random.

Added functions interface that allows creating functions accepting string and unsigned integer and returning string

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added a function that removes randomly generated parameters from URL

aterehov committed:
Added removeGarbageParametersFromURL function

Added removeGarbageParameterFromURL function that removes parameters of some minimal length that look random
Added functions interface that allows creating functions accepting string and unsigned integer and returning string

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=49305) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=49305) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=49305) it.</sub>

aterehov committed:
Merge branch 'ClickHouse:master' into master

aterehov committed:
Fixed function interface in order to use last version of DB::Exception

And removed trailing whitespaces

rschu1ze reviewed (commented):
@aterehov Thanks for this contribution. This PR would need
- a test in tests/queries/0_stateless (see [here](https://clickhouse.com/docs/en/development/tests))
- documentation [here](https://clickhouse.com/docs/en/sql-reference/functions/url-functions)

aterehov committed:
Added documentation

Fixed whitespaces

robot-clickhouse commented:
<!-- automatic status comment for PR #49305 from aterehov/ClickHouse:master -->
This is an automated comment for commit 4f2b452d6992cbb3d11ce714283a5668cccd8623 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49305/4f2b452d6992cbb3d11ce714283a5668cccd8623/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🟡 pending</td></tr>
</table>

aterehov committed:
Made code more cross-platform

Replaced malloc.h with stdlib.h
Replaced some C libraries with C++ alternatives
Replaced long integers with int64_t

aterehov committed:
Replaced stdlib.h with cstdlib

rschu1ze commented on src/Functions/URL/removeGarbageParametersFromURL.cpp:
```diff
@@ -0,0 +1,427 @@
+        size_t cur_offset = offsets[0];
+        const char * pdata = reinterpret_cast<const char *>(&data[prev_offset]);
+        const char * pdata_end = reinterpret_cast<const char *>(&data[cur_offset]) - 1;
+        size_t size = pdata_end - pdata;
```
`url_length`

rschu1ze commented on src/Functions/URL/removeGarbageParametersFromURL.cpp:
```diff
@@ -0,0 +1,427 @@
+        const char * pdata = reinterpret_cast<const char *>(&data[prev_offset]);
+        const char * pdata_end = reinterpret_cast<const char *>(&data[cur_offset]) - 1;
+        size_t size = pdata_end - pdata;
+        const char * read = pdata;
```
`cur` is (as far as I see) the most common term for the current read pointer

rschu1ze commented on src/Functions/URL/removeGarbageParametersFromURL.cpp:
```diff
@@ -0,0 +1,427 @@
+        size_t size = pdata_end - pdata;
+        const char * read = pdata;
+
+        /// Calculate length of webpage address and copy it
```
Maybe `Calculate length of the protocol, domain and path`

rschu1ze commented on src/Functions/URL/removeGarbageParametersFromURL.cpp:
```diff
@@ -0,0 +1,427 @@
+                url_size = i;
+                break;
+            }
+            else if (read[i] == 0)
```
Which situation does this handle? Is `0` in an URL valid?

rschu1ze commented on src/Functions/URL/removeGarbageParametersFromURL.cpp:
```diff
@@ -0,0 +1,427 @@
+    /// @param size length of data to be added
+    /// @param pdata pointer to the first character of data
+    /// @param res_offsets offsets of output string
+    static void copy(ColumnString::Chars & res_data, size_t size, const char * pdata, ColumnString::Offsets & res_offsets)
```
`copyIntoOutput`.

Also let's bring the parameters into a more logical order: `const char * from, size_t from_length, ColumnString::Chars & res_data, ColumnString::Offsets & res_offsets`

rschu1ze commented on src/Functions/URL/removeGarbageParametersFromURL.cpp:
```diff
@@ -0,0 +1,427 @@
+    /// @brief End the string with null terminator
+    /// @param res_data String to finalize
+    /// @param res_offsets Offsets of string
+    static void finalize_data(ColumnString::Chars & res_data, ColumnString::Offsets & res_offsets)
```
camelCase: `finalizeData`

rschu1ze commented on src/Functions/URL/removeGarbageParametersFromURL.cpp:
```diff
@@ -0,0 +1,427 @@
+        return !ret;
+    }
+
+    /// @brief Check if string is long enough and if it could be hex or base64 for url
```
Doxygen-style parameter docs are quite rare in the codebase. I would say if the parameters have expressive enough names then we can remove it and just keep l. 199.

rschu1ze commented on src/Functions/URL/removeGarbageParametersFromURL.cpp:
```diff
@@ -0,0 +1,427 @@
+                break;
+            }
+        }
+        if (!check_hex && !check_base64)
```
(cosmetic: let's omit {} in single-line if/for/while statements)

rschu1ze commented on src/Functions/URL/removeGarbageParametersFromURL.cpp:
```diff
@@ -0,0 +1,427 @@
+    {
+        /// Deciding how much intervals to split symbols into and whether the string is long enough to be handled
+        int64_t length = c_end - c_start;
+        int l = static_cast<int>(3.32 * log10(length)) + 1;
```
To a random future reader, this looks like too much magic. Could you please document the kind of test/algorithm used in `pearsonIsUniform()?

aterehov committed:
Fixed amount of freedom for chi-square distribution

Removed unused cstring library

aterehov committed:
Improved code
Fixed in-source documentation
Made function interface more abstract
Added stateless functional tests
Made names more speaking
Made function names more speaking
Fixed codestyle
Added support for multiple URL requests
Added info about Pearson's chi-squared test

aterehov committed:
Added documentation
Removed unused external link ILLEGAL_TYPE_OF_ARGUMENT in function interface

rschu1ze commented:
@aterehov Thanks for updating this PR and your effort! This seems a student/coursework project if I understand correctly.

To be honest, I am hesitant to merge this PR. Not because it has bad quality but because I don't really see practical use cases for the new functionality. 1. Do random URL parameters really occur in practice? It is very unlikely that parameters were garbled during transmission. I can only imagine that a malicious attacker constructs garbage or an accidental user input error. 2. And if such parameters occur, does the result of a magic repair function (which can and does only work probabilistically) produce a URL that still bears useful semantics for the code / program interpreting this URL? There is a principle [garbage in, garbage out](https://en.wikipedia.org/wiki/Garbage_in,_garbage_out) which in my view also applies to URLs. Any attempt to repair them is basically doomed to fail in the general case.

aterehov committed:
Merge branch 'ClickHouse:master' into master

aterehov committed:
Fixed in-source documentation
Excluded removeGarbageParametersFromURL from docs spell check

aterehov committed:
Fixed tests
Added temporary hits table with test data

aterehov committed:
Fixed tests
Added ordering to test.hits table

aterehov committed:
Fixed test

aterehov committed:
Fixed tests

aterehov committed:
Fixed tests

aterehov committed:
Disabled stress tests
Fixed potential undefined behaviours

aterehov committed:
Disabled parallel tests

aterehov committed:
Bypassed Clang Tidy warning about uninitialized values in allocated memory
Enabled stress tests

aterehov committed:
Bypassed Clang Tidy warning about uninitialized values in allocated memory

aterehov commented:
@rschu1ze Yes, this used to be a coursework project when I started doing it. Later I decided to update the function so it could be merged into ClickHouse. 

The function seems to be finished now, and the errors in checks do not seem to be related to removeGarbageParametersFromURL. You can continue working on it and merge it into ClickHouse if you need it one day. 
As for me, I need to make more efforts with my studies at the moment, so it is unlikely I will return to this pull request in the future.

rschu1ze commented:
@aterehov Appreciated it, thanks for the update! As mentioned above, I don't see an urgent need to have such a function in ClickHouse at the moment. For that reason, I'll close the PR for now. If someone has a different opinion, please let me know, we can reopen it then.

rschu1ze commented:
Lol, I just noticed the feature was motivated by #6498. Okay then, I'll continue/merge it if I find some time.

rschu1ze commented:
We can continue this another time.
